pallets/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,1861 @@
+			asset_pair: AssetPair,
+			loyalty_curve: Option<LoyaltyCurve>,
+		},
+
+		/// Farm was destroyed.
+		FarmDestroyed { id: PoolId, who: AccountIdOf<T> },
+
+		/// New LP tokens was deposited.
+		SharesDeposited {
+			farm_id: GlobalPoolId,
+			liq_pool_farm_id: PoolId,
+			who: AccountIdOf<T>,
+			amount: Balance,
+			lp_token: T::CurrencyId,
+			nft_class_id: NftClassIdOf<T>,
+			nft_instance_id: NftInstanceIdOf<T>,
+		},
+
+		/// Rewards was claimed.
+		RewardClaimed {
+			farm_id: GlobalPoolId,
+			liq_pool_farm_id: PoolId,
+			who: AccountIdOf<T>,
+			claimed: Balance,
+			reward_currency: T::CurrencyId,
+		},
+
+		/// LP tokens was withdrawn.
+		SharesWithdrawn {
+			farm_id: GlobalPoolId,
+			liq_pool_farm_id: PoolId,
+			who: AccountIdOf<T>,
+			lp_token: T::CurrencyId,
+			amount: Balance,
+		},
+
+		/// Liquidity mining for asset pair was canceled.
+		LiquidityMiningCanceled {
+			farm_id: GlobalPoolId,
+			liq_pool_farm_id: PoolId,
+			who: AccountIdOf<T>,
+			asset_pair: AssetPair,
+		},
+
+		/// Liquidity mining for asset pair was resumed.
+		LiquidityMiningResumed {
+			farm_id: GlobalPoolId,
+			liq_pool_farm_id: PoolId,
+			who: AccountIdOf<T>,
+			asset_pair: AssetPair,
+			multiplier: PoolMultiplier,
+		},
+
+		/// Liquidity pool was removed from farm.
+		LiquidityPoolRemoved {
+			farm_id: GlobalPoolId,
+			liq_pool_farm_id: PoolId,
+			who: AccountIdOf<T>,
+			asset_pair: AssetPair,
+		},
+
+		/// Undistributed rewards was withdrawn from farm.
+		UndistributedRewardsWithdrawn {
+			farm_id: GlobalPoolId,
+			who: AccountIdOf<T>,
+			reward_currency: T::CurrencyId,
+			amount: Balance,
+		},
+
+		/// Liquidity pool multiplier was updated.
+		LiquidityPoolUpdated {
+			farm_id: GlobalPoolId,
+			liq_pool_farm_id: PoolId,
+			who: AccountIdOf<T>,
+			asset_pair: AssetPair,
+			multiplier: PoolMultiplier,
+		},
+
+		/// Farm's(`GlobalPool`) accumulated reward per share was updated.
+		FarmAccRPZUpdated {
+			farm_id: GlobalPoolId,
+			accumulated_rpz: Balance,
+			total_shares_z: Balance,
+		},
+
+		/// Liquidity pool's `accumulated_rpvs` was updated.
+		LiquidityPoolAccRPVSUpdated {
+			farm_id: GlobalPoolId,
+			liq_pool_farm_id: PoolId,
+			accumulated_rpvs: Balance,
+			total_valued_shares: Balance,
+		},
+	}
+
+	/// Id sequencer for `GlobalPool` and `LiquidityPoolYieldFarm`.
+	#[pallet::storage]
+	#[pallet::getter(fn pool_id)]
+	pub type PoolIdSequencer<T: Config> = StorageValue<_, PoolId, ValueQuery>;
+
+	/// Sequencer for nft part of nft id.
+	#[pallet::storage]
+	pub type NftInstanceSequencer<T: Config> = StorageValue<_, primitives::InstanceId, ValueQuery>;
+
+	/// Global pool details.
+	#[pallet::storage]
+	#[pallet::getter(fn global_pool)]
+	type GlobalPoolData<T: Config> = StorageMap<_, Twox64Concat, GlobalPoolId, GlobalPool<T>, OptionQuery>;
+
+	/// Liquidity pool yield farm details.
+	#[pallet::storage]
+	#[pallet::getter(fn liquidity_pool)]
+	type LiquidityPoolData<T: Config> = StorageDoubleMap<
```

Comment:
I would personally set some constants for max amount of pools (any map really) and then benchmark against that max number since maps/double maps might take longer to read/write when there is a lot of entries. Something like 255 should be sufficient.